jobrunner/run.py

```diff
@@ -99,6 +111,44 @@ def handle_pending_job(job):
                 mark_job_as_running(job)
 
 
+def handle_pending_job_api(job, api):
+    if job.cancelled:
+        # Mark the job as running and then immediately invoke
+        # `handle_running_job` to deal with the cancellation. This slightly
+        # counterintuitive appraoch allows us to keep a simple, consistent set
+        # of state transitions and to consolidate all the kill/cleanup code
+        # together. It also means that there aren't edge cases where we could
+        # lose track of jobs completely after losing database state
+        mark_job_as_running(job)
+        handle_running_job_api(job, api)
+        return
+
+    awaited_states = get_states_of_awaited_jobs(job)
+    if State.FAILED in awaited_states:
+        mark_job_as_failed(
+            job, "Not starting as dependency failed", code=StatusCode.DEPENDENCY_FAILED
+        )
+    elif any(state != State.SUCCEEDED for state in awaited_states):
+        set_message(
+            job, "Waiting on dependencies", code=StatusCode.WAITING_ON_DEPENDENCIES
+        )
+    else:
+        not_started_reason = get_reason_job_not_started(job)
+        if not_started_reason:
+            set_message(job, not_started_reason, code=StatusCode.WAITING_ON_WORKERS)
+        else:
+            try:
+                set_message(job, "Preparing")
+                api.run(job_to_job_definition(job))
+            except JobError as exception:
+                mark_job_as_failed(job, exception)
+            except Exception:
+                mark_job_as_failed(job, "Internal error when starting job")
+                raise
+            else:
+                mark_job_as_running(job)
```

Comment:
My preference would be to put this in the `try` block along with the rest of the happy path.

Reply:
Yeah, I was just preserving the original function structure, but I agree

Reply:
I don't particularly care about this at all, but to my mind the original structure was correct. Exceptions raised by `mark_job_as_running` are from a different layer than those raised by "start_job" and shouldn't be handled by the final `except Exception` block - it wasn't the job the failed here, it was the job-runner.

Reply:
I think you're right. I'm afraid the case where the `mark_*()` function raises an exception didn't occur to me.